File: Code/folding.py
import numpy as np
import math
import matplotlib.pyplot as plt
import sys
from PIL import Image


#loading the binary file data----------------------------------------------------

timeseries_path= sys.argv[1]
f= open(timeseries_path, "rb")
time_series= np.fromfile(f, dtype= 'float32')
f.close()

#display_window = 20000 #no. of samples
#plt.plot(time_series[:display_window])
#plt.savefig("time_series.png")
#plt.close()	

#Folding the time series---------------------------------------------------------

#info needed for folding
phase_offset=-1 #default value if 3rd argument isnt given
P = float(sys.argv[2]) #in ms
tau = float(sys.argv[3]) #in ms
if len(sys.argv)==5:
	phase_offset = float(sys.argv[4])

num_bins = int(P/tau)					#native resolution	
bin_interval = P/num_bins   				 #time span of each bin

bins = [[] for i in range(num_bins)]			 #this array will contains the bins and their contents. Each row is represented by one bin

folded= np.zeros(num_bins)				#this array will contain the average value for each bin

bin_count=np.zeros(num_bins,dtype=int)  		 #this keep track of how many samples we have added to each bin

	
#For every sample, we have to decide which bin it goes to.
for i in range(len(time_series)):
	bin_index = round(((i*tau)%P)/bin_interval)   #this line decides which bin gets the ith sample
	
	if bin_index == num_bins: bin_index=0		
	bins[bin_index].append(time_series[i])
	bin_count[bin_index]+=1

bins_arr=np.array(bins, dtype=object)
for i in range(num_bins):
	folded[i]=np.sum(bins_arr[i])/bin_count[i]	#sum all values of each bin and divide by number of samples added to each

#print(folded)
if phase_offset != -1: #adjusting for the phase_offset in the folded profile
	bin_offset= round(num_bins*phase_offset)
	temp=np.copy(folded)
	temp[bin_offset:]= folded[:num_bins-bin_offset]
	temp[:bin_offset] = folded[num_bins-bin_offset:]
	folded=np.copy(temp)
	
#print(len(folded))
plt.plot(folded, marker='|')
plt.savefig("folded.png")
f= open("folded.txt", 'w')
np.savetxt(f,folded, fmt='%.18e')

im= Image.open("folded.png")
im.show()

File: Meetings/meeting 13 (1.8.21)/folding.c
#include<stdio.h>
#include<stdlib.h>
#include<string.h>
#include<math.h>

int main(int argc, char* argv[])
{
	char *timeseries_path=(char*)malloc(1*strlen(argv[1]));
	FILE *f;
	int N=0,i=0;
	float val, *time_series= (float*) malloc(sizeof(float)*1);
	long int totalsize;

	float phase_offset=-1.0, *bins=(float*) malloc(sizeof(float)*1);
	double P,tau,bin_interval;
	int num_bins,bin_index, *bin_count=(int*) malloc(sizeof(int)*1);
	int buffer_binsize;

	float *folded=(float*) malloc(sizeof(float)*1),*temp=(float*) malloc(sizeof(float)*1),sum;
	int bin_offset;

	FILE *gnuplot = popen("gnuplot -persistent", "w");
	FILE *foldedfile = fopen("folded.txt", "w");
	int numcommands= 4;
	//char *gnuplotcommands[] = {"set title \"Folded\";set terminal png size 500,400;set output 'folded.png';plot \"folded.txt\" u 1:2 w lp"};
	char *gnuplotcommands[] = {"set title \"Folded\"","set terminal png size 800,600", "set output 'folded.png'" ,"plot \"folded.txt\" u 1:2 w lp"};


//loading the binary file data----------------------------------------------------

	strcpy(timeseries_path,argv[1]);
	f= fopen(timeseries_path,"rb");

	//calculating total number of samples N	
	fseek(f,0,SEEK_END);
	totalsize= ftell(f);
	N= totalsize/sizeof(float);
	
	//putting data in time_series
	fseek(f,0,SEEK_SET);
	time_series=(float*)realloc(time_series,sizeof(float)*N);
	for(i=0; i<N ;i++)
	{
		fread(&val, sizeof(float), 1, f);
		time_series[i]=val;
	}

	fclose(f);

//Folding the time series---------------------------------------------------------

//info needed for folding

	//commandline arguments	
	P = strtod(argv[2],NULL);
	tau=strtod(argv[3],NULL);
	if(argc==5) phase_offset=strtof(argv[4],NULL);

	
	num_bins=(int)P/tau;
	bin_interval= (double)P/num_bins;
	bin_count=(int*)realloc(bin_count,num_bins*sizeof(int));
	for(i=0;i<num_bins;i++) bin_count[i]=0;
	
	//buufer_binsize keeps extra space for each bin in the 'bins' array
	buffer_binsize= (int) (N*1.1)/num_bins;
	bins=(float*) realloc(bins, num_bins*buffer_binsize*sizeof(float)); 
	for(i=0;i<N;i++) bins[i]=0.0;

	//the actual folding loop
	for(i=0;i<N;i++)
	{
	bin_index =(int)round((fmod(i*tau,P))/bin_interval); //decides which bin the ith sample goes to
	if (bin_index == num_bins) bin_index=0;
	bin_count[bin_index]+=1;
	
	bins[bin_index*buffer_binsize+ bin_count[bin_index]-1]=time_series[i];
	}

	//summing all values in each bin and taking average to get folded profile array	
	folded= (float*)realloc(folded,num_bins*sizeof(float));
	for(i=0;i<num_bins; i++) folded[i]=0.0;

	for(i=0;i<num_bins;i++)
	{
		sum=0.0;
		for(int j=0;j<bin_count[i];j++) sum+=bins[i*buffer_binsize+j];
		folded[i]=sum/bin_count[i];
	}

	//adjusting for the phase_offset in the folded profile
	if (phase_offset != -1)
	{
		bin_offset= (int)round(num_bins*phase_offset);
		temp= (float*) realloc(temp, num_bins*sizeof(float));

		for(i=0;i<num_bins - bin_offset; i++) temp[bin_offset + i] = folded[i];
		for(i=0;i<bin_offset; i++) temp[i] = folded[num_bins- bin_offset+i]; 
		folded=(float*) realloc(temp, num_bins*sizeof(float));	
	}	

	//saving the folded profile in a textfile

	char header[500];
	sprintf(header, "#Path                             :%s\n#Period                           :%lf\n#Sampling Interval                :%lf\n#phase_offset                     :%f\n#Total no. of lines is 5(including this line) and the numeric values at each line start from the 36th character\n",timeseries_path,P,tau,phase_offset);
	fprintf(foldedfile, header);
	
	for(i=0;i<num_bins;i++)
	{
	fprintf(foldedfile, "%f %f\n", i/((float)num_bins), folded[i]);
	}
	
	// automatically plotting the folded profile using gnuplot
	for(i=0;i<numcommands ;i++)
	{
	fprintf(gnuplot, "%s\n", gnuplotcommands[i]);
	}
		

	return 0;
}
